logging/WFCORE-4807-log-routing.adoc

```diff
@@ -0,0 +1,106 @@
+= Log Routing
+:author:            James R. Perkins
+:email:             jperkins@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Currently any modules that log generally log onto the system log context. In some cases, for example a deployment using
+a logging profile, the loggers may end up getting created on the wrong `LogContext`. For example if `a.war` and `b.war`
+both depend on module `org.jboss.custom.module` and at least one deployment uses a logging profile whichever deployment
+initializes the custom module first will have the loggers created on that `LogContext`. If a logging profile is being
+used it would be that profiles log context.
+
+https://issues.jboss.org/browse/LOGMR-264[LOGMR-264] will introduce a way to route log messages based on the log context.
+This would allow messages from modules to be routed to deployment specific log contexts. It would also introduce a
+change in behavior which is the reason for an RFE. A new attribute should be introduce to disable the new routing and
+allow the old behavior.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-4897[WFCORE-4807]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1420[EAP7-1420]
+* https://issues.jboss.org/browse/LOGMR-264[LOGMR-264]
+* https://issues.redhat.com/browse/WFCORE-3376[WFCORE-3376]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* TBD
+
+=== Testing By
+
+* [x] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+* WildFly Core
+* WildFly
+* JBoss Log Manager
+
+=== Other Interested Projects
+
+* Any consumer of WildFly Core that uses the logging subsystem.
+
+== Requirements
+
+=== Hard Requirements
+
+* The attribute will be a simple boolean attribute which determines if the new or old behavior will be used.
+* When the new behavior is disabled the behavior should be the same as it currently works.
+    ** This means that when disabled the bug in https://issues.redhat.com/browse/WFCORE-3376[WFCORE-3376] would
+       still be present.
+* The `LogContext` should be associated with a deployments class loader.
+    ** This means that if the TCCL is set to the deployment class loader log messages will be routed to the deployments
+       log context.
+    ** That may include messages that previously went to the `server.log` could end up in a handler defined on the
```

Comment:
What is the impact on messages related to deploying and undeploying the deployment?

Reply:
If the TCCL is set to the deployments class loader then messages get routed to the deployments log context. For example:
```
09:54:42,716 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 85) WFLYUT0021: Registered web context: '/test-logging-1' for server 'default-server'
```
gets logged to the deployments log context.

But the following is logged in the servers log configuration:
```
2020-01-24 09:54:42,301 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-6) WFLYSRV0027: Starting deployment of "test-logging-1.war" (runtime-name: "test-logging-1.war")
2020-01-24 09:54:42,437 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-6) WFLYSRV0018: Deployment "deployment.test-logging-1.war" is using a private module ("org.wildfly.security.manager") which may be changed or removed in future versions without notice.
2020-01-24 09:54:42,611 INFO  [org.jboss.as.arquillian] (MSC service thread 1-6) Arquillian deployment detected: ArquillianConfig[service=jboss.arquillian.config."test-logging-1.war",unit=test-logging-1.war,tests=[org.jboss.as.test.integration.logging.LoggingProfileModuleTestCase]]
2020-01-24 09:54:42,748 INFO  [org.jboss.as.server] (management-handler-thread - 1) WFLYSRV0010: Deployed "test-logging-1.war" (runtime-name : "test-logging-1.war")
2020-01-24 09:54:43,362 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-2) WFLYSRV0028: Stopped deployment test-logging-1.war (runtime-name: test-logging-1.war) in 42ms
2020-01-24 09:54:43,379 INFO  [org.jboss.as.repository] (management-handler-thread - 1) WFLYDR0002: Content removed from location /home/jperkins/projects/wildfly/wildfly/testsuite/integration/basic/target/wildfly/standalone/data/content/56/6daf8be2c6588d53cf5a1acda67e2745821633/content
2020-01-24 09:54:43,380 INFO  [org.jboss.as.server] (management-handler-thread - 1) WFLYSRV0009: Undeployed "test-logging-1.war" (runtime-name: "test-logging-1.war")
```

Reply:
Clicked comment too soon :)

Basically the idea is, and maybe I should add this somewhere in the doc, is that stuff coming from say Weld or Hibernate that is associated with the deployment should end up in the deployments logs. Basically any messages that come from server dependencies where the TCCL is set to the deployments class loader should end up in the applications logs.

Reply:
That might be a bit odd. I don't know if anything can be done about it. But I doubt that much thought is given to what TCCL is in place when logging calls get added. That "Registered web context" message is more of a server.log message than the "Deployment "deployment.test-logging-1.war" is using a private module" one is. (Both are valid server.log messages; the 'private module' one is just more a detail about a particular deployment than the 'Registered web context' is.

Reply:
I can definitely see your point. The problem we're attempting to solve is that modules used by deployments may create loggers on the wrong log context. For example given two deployments, `a.war` and `b.war`, that each have their own log context and both have a dependency on `org.jboss.example.module` have a race to see which log context creates the static loggers associated with the module.

In other words; if `b.war` first initializes the loggers in the `org.jboss.example.module` then all log messages, even if they are initiated by `a.war`, from the module end up in the `b.war` log context. Note this assumes that `b.war` does use either a logger profile or have a logging configuration file in it that is configured by a DUP in the logging subsystem.

The two solutions would be:

1. All messages from modules get logged on the server level log context
2. Messages from modules that seemingly belong to the deployment get logged with the deployment. Seemingly here means the TCCL is set to the deployments class loader.

The problem with 1 is there is no real way to know that the logger is being created on a module and not a deployment. Short of inspecting the class loader and assuming that anything who's module name that does *not* begin with "deployment" is global module.

The problem with 2, as you point out, is it's a guess too at which log messages are said to be "deployment" log messages. Checking the TCCL is the deployments class loader is how the assumption is made and seems somewhat safe. Though I do agree and can see this could be considered wrong. This was part of the reason for the RFE to turn this off.

It seems the options are:

1. Reject the RFE and live with the bug
2. Have the flag like this document proposes which we could change to opt-in instead of opt-out.
3. We change the RFE to be a flag that would inspect the `ClassLoader` of the caller class for a module name of "deployment.*". Possibly even use a user defined pattern, but I'm not too sure I love that idea :)

I'm definitely open for input here. I can say since JBoss AS 7 users have wanted something akin to the `TCCLFilter` though admittedly in recent years this has not come up as much. Also I'm making some assumptions on how the `TCCLFilter` actually worked too. I've only heard of it before :)

Reply:
@jamezp 

What does this look like if something like kitchensink.war gets deployed? In the discussion above there were two "odd" (i.e. worth discussing) messages. With complex deployments are there a lot more?

Is it correct to say that in a case where we really want something to go to server.log but it ends up not, that we could wrap the log call in a TCCL set/restore block? So we have a fix for those. If something still ended up in the server.log, well maybe we could also do a TCCL set/restore block, if the relevant code had access to the deployment's CL.

Reply:
@bstansberry Here's the logs from the deploy operation. The `[app-log]` is from the `kitchensink.war`.
```
11:06:34,392 INFO  [org.jboss.as.repository] (management-handler-thread - 1) WFLYDR0001: Content added at location /home/jperkins/projects/wildfly/wildfly/build/target/wildfly-22.0.0.Beta1-SNAPSHOT/standalone/data/content/a8/58ea34cd6065235347abf41b90a6e2f0c07065/content
11:06:34,408 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-7) WFLYSRV0027: Starting deployment of "kitchensink.war" (runtime-name: "kitchensink.war")
11:06:34,737 INFO  [org.jboss.as.jpa] (MSC service thread 1-4) WFLYJPA0002: Read persistence.xml for primary
11:06:34,745 WARN  [org.jboss.as.connector] (MSC service thread 1-4) WFLYJCA0091: -ds.xml file deployments are deprecated. Support may be removed in a future version.
11:06:34,864 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-1) WFLYJCA0001: Bound data source [java:jboss/datasources/KitchensinkQuickstartDS]
11:06:34,892 INFO  [org.jboss.weld.deployer] (MSC service thread 1-8) WFLYWELD0003: Processing weld deployment kitchensink.war
11:06:34,988 INFO  [org.hibernate.validator.internal.util.Version] (MSC service thread 1-8) HV000001: Hibernate Validator 6.0.21.Final
11:06:35,107 INFO  [org.jboss.as.ejb3.deployment] (MSC service thread 1-8) WFLYEJB0473: JNDI bindings for session bean named 'MemberRegistration' in deployment unit 'deployment "kitchensink.war"' are as follows:

	java:global/kitchensink/MemberRegistration!org.jboss.as.quickstarts.kitchensink.service.MemberRegistration
	java:app/kitchensink/MemberRegistration!org.jboss.as.quickstarts.kitchensink.service.MemberRegistration
	java:module/MemberRegistration!org.jboss.as.quickstarts.kitchensink.service.MemberRegistration
	java:global/kitchensink/MemberRegistration
	java:app/kitchensink/MemberRegistration
	java:module/MemberRegistration

11:06:35,108 INFO  [org.infinispan.CONTAINER] (ServerService Thread Pool -- 78) ISPN000128: Infinispan version: Infinispan 'Corona Extra' 11.0.4.Final
11:06:35,136 INFO  [org.jboss.as.jpa] (ServerService Thread Pool -- 78) WFLYJPA0010: Starting Persistence Unit (phase 1 of 2) Service 'kitchensink.war#primary'
11:06:35,149 INFO  [org.hibernate.jpa.internal.util.LogHelper] (ServerService Thread Pool -- 78) HHH000204: Processing PersistenceUnitInfo [
	name: primary
	...]
11:06:35,168 INFO  [org.infinispan.CONFIG] (MSC service thread 1-3) ISPN000152: Passivation configured without an eviction policy being selected. Only manually evicted entities will be passivated.
11:06:35,169 INFO  [org.infinispan.CONFIG] (MSC service thread 1-3) ISPN000152: Passivation configured without an eviction policy being selected. Only manually evicted entities will be passivated.
11:06:35,214 INFO  [io.jaegertracing.internal.JaegerTracer] (MSC service thread 1-8) No shutdown hook registered: Please call close() manually on application shutdown.
11:06:35,215 INFO  [org.hibernate.Version] (ServerService Thread Pool -- 78) HHH000412: Hibernate Core {5.3.20.Final}
11:06:35,216 INFO  [org.hibernate.cfg.Environment] (ServerService Thread Pool -- 78) HHH000206: hibernate.properties not found
11:06:35,241 INFO  [org.infinispan.PERSISTENCE] (ServerService Thread Pool -- 16) ISPN000556: Starting user marshaller 'org.wildfly.clustering.infinispan.spi.marshalling.InfinispanProtoStreamMarshaller'
11:06:35,320 INFO  [org.hibernate.annotations.common.Version] (ServerService Thread Pool -- 78) HCANN000001: Hibernate Commons Annotations {5.0.5.Final}
11:06:35,329 INFO  [org.jboss.weld.Version] (MSC service thread 1-3) WELD-000900: 3.1.5 (Final)
11:06:35,378 INFO  [org.jboss.as.clustering.infinispan] (ServerService Thread Pool -- 16) WFLYCLINF0002: Started http-remoting-connector cache from ejb container
[app-log] 11:06:35,476 INFO  [org.jboss.as.jpa] (ServerService Thread Pool -- 16) WFLYJPA0010: Starting Persistence Unit (phase 2 of 2) Service 'kitchensink.war#primary'
[app-log] 11:06:35,597 INFO  [org.hibernate.dialect.Dialect] (ServerService Thread Pool -- 16) HHH000400: Using dialect: org.hibernate.dialect.H2Dialect
[app-log] 11:06:35,601 WARN  [org.hibernate.dialect.H2Dialect] (ServerService Thread Pool -- 16) HHH000431: Unable to determine H2 database version, certain features may not work
[app-log] 11:06:35,645 INFO  [org.hibernate.envers.boot.internal.EnversServiceImpl] (ServerService Thread Pool -- 16) Envers integration enabled? : true
[app-log] 11:06:36,030 INFO  [org.hibernate.tool.schema.internal.SchemaCreatorImpl] (ServerService Thread Pool -- 16) HHH000476: Executing import script 'ScriptSourceInputFromUrl(vfs:/content/kitchensink.war/WEB-INF/classes/import.sql)'
[app-log] 11:06:36,104 INFO  [io.smallrye.metrics] (MSC service thread 1-7) MicroProfile: Metrics activated (SmallRye Metrics version: 2.4.4)
[app-log] 11:06:36,141 WARN  [org.jboss.weld.Bootstrap] (MSC service thread 1-7) WELD-000146: BeforeBeanDiscovery.addAnnotatedType(AnnotatedType<?>) used for class com.sun.faces.flow.FlowDiscoveryCDIHelper is deprecated from CDI 1.1!
[app-log] 11:06:36,503 INFO  [javax.enterprise.resource.webcontainer.jsf.config] (ServerService Thread Pool -- 16) jsf.config.listener.version
[app-log] 11:06:37,067 INFO  [org.jboss.resteasy.resteasy_jaxrs.i18n] (ServerService Thread Pool -- 16) RESTEASY002225: Deploying javax.ws.rs.core.Application: class org.jboss.as.quickstarts.kitchensink.rest.JaxRsActivator$Proxy$_$$_WeldClientProxy
[app-log] 11:06:37,116 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 16) WFLYUT0021: Registered web context: '/kitchensink' for server 'default-server'
11:06:37,178 INFO  [org.jboss.as.server] (management-handler-thread - 1) WFLYSRV0010: Deployed "kitchensink.war" (runtime-name : "kitchensink.war")
```

> Is it correct to say that in a case where we really want something to go to server.log but it ends up not, that we could wrap the log call in a TCCL set/restore block? So we have a fix for those. If something still ended up in the server.log, well maybe we could also do a TCCL set/restore block, if the relevant code had access to the deployment's CL.

Yes that is correct. We could wrap the TCCL to ensure the log ends up being logged to the correct context.

Reply:
Coming back to this. @kabir gave me a good idea of adding an annotation to automate the wrapping. I created a `@ThreadContext`, maybe I'll use a better name, annotation which sets and resets the TCCL based on the type passed in or `getClass().getClassLoader()`. I've updated all the branches and that seems to work quite well.